### ISSUE TITLE
My steering wheel doesn't work.

### ISSUE THREAD

**Issue by MRZA-MRZA** [virtual-programming/dirtshowdown-linux#38]
Hello. I connected my steering wheel controller, calibrated it and ran DiRT Showdown, but I didn't find how to adjust my wheel and I couldn't map any buttons. It works well in another game.
Here is full log:
http://pastebin.com/aX1VPKW2

**Comment by vpltd-kgalaj** [comment]
Later on, in the log you pasted, there are lines:

> [thread 00000034][E][6672]: Gamepads: Device couldn't be activated, because no SDL mapping for device GUID 03000000f806000003a3000000010000 has been found!
> [thread 00000034][I][6672]: Gamepads:     To activate the device, please provide the appropriate mapping in SDL2 format through command-line ( --eon_gamepad_mapping= option ),
> [thread 00000034][I][6672]: Gamepads:     SDL_GAMECONTROLLERCONFIG environment variable, or SDLGamepad.config file, located in game data folder.
> [thread 00000034][I][6672]: Gamepads:     SDL2 format mapping strings can be determined using Steam Big Picture configurator.

It's not named by name, but it's the GUID of your gamepad.

Generally, our games recognise gamepads with known SDL mapping, when they're provided to the game through command-line (for example entered in Steam client), through SDL_GAMECONTROLLERCONFIG environment variable (this is the way Steam Big Picture gamepad configuration is passed on to the game), or through entering the gamepad ID and the settings in SDLGamepad.config file, provided in the game's data folder. It's possible your joystick is a model which we haven't encountered yet, and so we have no mapping for it.

If you launch Steam Big Picture, configure your wheel's mapping in it, and then launch the game from Big Picture, it should pass the configuration on to the game.

If you prefer the game to know about your wheel model directly, follow these steps.

I. Making sure that Steam generates a controller mapping for you, and getting it.

1. Launch Steam.
2. Connect your gamepad and make sure it's on.
3. Launch Steam Big Picture, if it's not in Big Picture mode (it's an icon depicting a gamepad in the top right corner of Steam window).
4. When Steam's Big Picture menu shows up, navigate to Settings (it's an icon near the 'exit' one in top right corner of the screen.
5. Select 'Controller' option.
6. At this point you should see your controller mappings. Make sure that they're all as you want them, and then select 'Save' option.
7. Steam may ask you to share your mapping with others at this point - agree or decline, it shouldn't matter.
8. Exit Steam.
9. Navigate to folder '~/.local/shared/Steam'.
10. Find file config.vdf there.
11. In it, look for a line of text that starts with "SDL_GamepadBind". The rest of the line would be a jumble of letters and numbers looking a bit like this (example): "5e040000000000008e02000000000000,Wireless 360 Controller,a:b0,b:b1,y:b3,x:b2,start:b10,guide:b8,back:b9,leftstick:b6,rightstick:b7,leftshoulder:b4,rightshoulder:b5,dpup:b11,dpleft:b13,dpdown:b12,dpright:b14,leftx:a0,lefty:a1,rightx:a3,righty:a4,lefttrigger:a2,righttrigger:a5,".
12. If there is such a line, the exact text between quotation marks in the rest of the line, excluding the quotation marks, is your controller mapping. Copy it somewhere.
13. Close the file. Don't save, especially if you made any changes, as this could potentially break your Steam client if you did.
14. If there is no such line in the file, go back to step 1, and get through steps, changing something in controller mappings in step 6, and select Save again to force Steam to save the mapping. Then get through remaining steps, open the config.vdf file again, and look for the controller mapping again, until you have it.

II. Making sure the game knows the mapping

Look for SDLGamepad.config file in game's data folder in Steam (may not apply to all games, there's also command-line option). Open it, add your game controller ID and mapping to it, and save it. Example mapping for example controler ID:
----------------------
5e040000000000008e02000000000000,Wireless 360 Controller,a:b0,b:b1,y:b3,x:b2,start:b10,guide:b8,back:b9,leftstick:b6,rightstick:b7,leftshoulder:b4,rightshoulder:b5,dpup:b11,dpleft:b13,dpdown:b12,dpright:b14,leftx:a0,lefty:a1,rightx:a3,righty:a4,lefttrigger:a2,righttrigger:a5,
----------------------

At this point, if you launch the game again, it should recognize your gamepad correctly. You will still need to redefine its controls in-game, but only once, after that it should remember them.

If you feel that a gamepad is recognized incorrectly in one of our games, or some controls aren't usable, you can do the same thing to override our mapping with the one defined using Steam's Big Picture.

III. (optional) Getting the controller mapping to us

If you send us the text of the mapping you arrived at, for example in your reply here, we will include it in the mappings we include predefined with all our future games and updates, and this will make the games recognize your gamepad in the future without the need for such steps.

Please advise.

**Comment by MRZA-MRZA** [comment]
I don't have config.vdf in ~/.local/share/Steam/
There is config.vdf in ~/.local/share/Steam/config/ but it doesn't contain "SDL_GamepadBind"

**Comment by vpltd-kgalaj** [comment]
This may be the right file - the guide was made and last fack-checked two years ago and Steam gets continually updated. I just checked my file, and the SDL_GamepadBind item is there for me in config.vdf in config folder, as one of the last lines in the file, at the very end.

For SDL_GamepadBind to be there you need to have configured the gamepad in Steam Big Picture, assigning actions to its inputs at least once for it to show up. Did you do it? Possibly it now saves the new value as you exit Steam, not earlier.

**Comment by MRZA-MRZA** [comment]
I tried again. Saved configuration in Big Picture, exited Steam, checked config.vdf: no "SDL_GamepadBind"
Should I assign all buttons?

**Comment by vpltd-kgalaj** [comment]
No, it shouldn't make any difference, the config should also work with partial assignments.

Googling for SDL_GamepadBind provides a variety of sources that say to use it, and in none of these sources there is any update that would say that things stopped working like this. Even in sources like https://github.com/gabomdq/SDL_GameControllerDB , which relies on SDL_GamepadBind for all its operations. So if you don't have this variable in your config/config.vdf, it seems likely that your Steam setup is somehow rare and atypical. I have it, so I can't reproduce the issue, so don't really have any way to help. Possibly you can find some information about this variable, on or arriving at valid SDL gamepad configuration string directly through Google?

**Comment by MRZA-MRZA** [comment]
I tried several times and BINGO! I got the configuration. Now my Racing Wheel works. Although, not all buttons work (I probably misconfigured something in Big Picture).
But is it possible to make it more user friendly? In most games controller support is just plug & play.

**Comment by vpltd-kgalaj** [comment]
It's the matter of knowing a sensible configuration for given gamepad or wheel. We maintain a database of such configurations and add it to our games, and if you have one of the controllers for which we already know a configuration, it's pretty much plug & play as well. In this case, we didn't have the mapping, and there's no standard for these on Linux so we have to learn as we meet new gamepad models. If you advise the configuration string you used, we'll add it to database and the gamepad will be supported from then on in any of our new releases and updates.

**Comment by MRZA-MRZA** [comment]
Tried to play using Xbox 360 gamepad. It was perfect. Which racing wheel is recommended to play DiRT Showdown?

**Comment by chrisaq** [comment]
@MRZA-MRZA I use a logitech G27 which works fine, no force feedback though.
You also need to limit the range of the wheel though, I wrote a small script to handle that:
https://github.com/chrisaq/lt-tools
